libsolidity/analysis/ControlFlowRevertPruner.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/ControlFlowRevertPruner.h>
+
+#include <libsolutil/Algorithms.h>
+
+namespace solidity::frontend
+{
+
+void ControlFlowRevertPruner::run(ASTNode const& _astRoot)
+{
+	// Run first iteration without following recursive loops
+	_astRoot.accept(*this);
+
+	// In the second iteration we process anything left pending
+	removePendingPaths();
+}
+
+void ControlFlowRevertPruner::removePendingPaths()
+{
+	// Anything still left pending must be a infinite recursive loop.
+	for (auto& [key, val]: m_functionReverts)
+	{
+		RevertState& revertState = std::get<State>(val);
+		auto functionFlow = m_cfg.functionFlow(
+			*std::get<Function>(key),
+			std::get<Contract>(key)
+		);
+
+		if (CFGNode* node = std::get<Node>(val))
+			for (auto const* functionCall: node->functionCalls)
+				switch (checkForReverts(*functionCall, std::get<Contract>(key)))
+				{
+				// Remove exits & add revert node
+				case RevertState::AllPathsRevert:
+					[[fallthrough]];
+				case RevertState::Pending:
+					node->exits.clear();
+					node->exits.push_back(functionFlow.revert);
+					break;
+				case RevertState::HasNonRevertingPath:
+					break;
+				}
+
+		if (revertState == RevertState::Pending)
+			revertState = RevertState::AllPathsRevert;
+	}
+}
+
+void ControlFlowRevertPruner::removeRevertingPaths(FunctionDefinition const& _function, ContractDefinition const* _contract)
+{
+	auto& functionFlow = m_cfg.functionFlow(_function, _contract);
```

Comment:
Please use `const&` if it is a `const&`. Also `FunctionFlow const& functionFlow` would not hurt.

Reply:
For free functions we want to pass `nullptr` and it felt wrong to use a reference for that even though it would work, technically. A reference suggests to me that it should never be null.. a pointer makes it clear that it could be null..

Reply:
I was just talking about the `auto&`.